src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -24605,6 +24605,1467 @@ void Init()
+
+    public string M(object? o)
+    {
+        if ((IsOk, o) is (true, null))
+        {
+            return Value; // incorrect warning
+        }
+
+        if (IsOk is true)
+        {
+            return Value;
+        }
+
+        throw null!;
+    }
+}
+", MemberNotNullWhenAttributeDefinition });
+
+            // We're not learning from tuple equality...
+            // Tracked by https://github.com/dotnet/roslyn/issues/50980
+            c.VerifyDiagnostics(
+                // (15,20): warning CS8603: Possible null reference return.
+                //             return Value; // incorrect warning
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "Value").WithLocation(15, 20)
+                );
+        }
+
+        [Fact, WorkItem(51020, "https://github.com/dotnet/roslyn/issues/51020")]
+        public void PatternOnTuple_IsExpression()
+        {
+            var src = @"
+#nullable enable
+public class C
+{
+    void M(object? o, object o2)
+    {
+        if ((o, o2) is (not null, null))
+        {
+            o.ToString();
+            o2.ToString(); // 1
+        }
+    }
+
+    void M2(object? o)
+    {
+        if (o is not null)
+        {
+            o.ToString();
+        }
+    }
+
+    void M3(object o2)
+    {
+        if (o2 is null)
+        {
+            o2.ToString(); // 2
+        }
+    }
+
+    void M4(object? o, object o2)
+    {
+        if ((true, (o, o2)) is (true, (not null, null)))
+        {
+            o.ToString(); // 3
+            o2.ToString();
+        }
+    }
+}
+";
+            // Note: we lose track in M4 because any learnings we make on elements of nested tuples
+            // apply to temps, rather than the original expressions/slots.
+            // This is unfortunate, but doesn't seem much of a priority to address.
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics(
+                // (10,13): warning CS8602: Dereference of a possibly null reference.
+                //             o2.ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "o2").WithLocation(10, 13),
+                // (26,13): warning CS8602: Dereference of a possibly null reference.
+                //             o2.ToString(); // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "o2").WithLocation(26, 13),
+                // (34,13): warning CS8602: Dereference of a possibly null reference.
+                //             o.ToString(); // 3
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "o").WithLocation(34, 13)
+                );
+        }
+
+        [Fact, WorkItem(51020, "https://github.com/dotnet/roslyn/issues/51020")]
+        public void PatternOnTuple_IsExpression_LongTuple()
+        {
+            var src = @"
+#nullable enable
+public class C
+{
+    void M(object? o, object o2)
+    {
+        if ((o2, o2, o2, o2, o2, o2, o2, o) is (null, null, null, null, null, null, null, not null))
```

Comment:
what happens when contradictory tests are performed on the same "input"? i.e.

```cs
    public void M(object? obj) {
        if ((obj, obj) is (null, not null)) { obj.ToString(); }
    }
```

Reply:
Haha, that's creative :-) I'll add the scenario. 
The last test (in our unspecified evaluation order) wins.

---
In reply to: [574028591](https://github.com/dotnet/roslyn/pull/51001#discussion_r574028591) [](ancestors = 574028591)